### ISSUE TITLE
How to handle equality of different representations of the same solution?

### ISSUE THREAD

**Issue by wouterkool** [PyVRP/PyVRP#311]
For undirected problems (CVRP with symmetric distances), we can have different representations of the same solution by reversing routes. Related to #101, which considers the distance between such instances, should we consider different 'representations' of the same/equivalent solutions equal in `Solution::operator==`? The challenge is that this depends on the problem instance which is not accessible in `Solution`.

Whether or not we view different representations of the same solution as equal or different affect how we maintain duplicates in the population: if we consider them different, we effectively may have many duplicates in the population, which may (or may not) affect performance.

**Comment by N-Wouda** [comment]
I'm OK with the current behaviour.

**Comment by leonlan** [comment]
> I just wanted to raise the discussion, I'm not saying that we should care that much. I think putting it into the diversity measure is a good idea, but as of now Population relies on the equality operator to remove duplicates, which we should then change? If so, doesn't it make sense to remove the equality operator alltogether, forcing an explicit choice of how to define equality when needed?

I don't think the current equality operator is wrong (except maybe for #315?). It's just that some problem variants have more symmetry resulting in solutions that are virtually the same for the problem at hand, but not in terms of the "equality" operator of the Solution data structure itself. Dealign with that symmetry can be delegated to the diversity measure/Population, since that's the only place it really matters.

**Comment by leonlan** [comment]
> I just wanted to raise the discussion, I'm not saying that we should care that much. I think putting it into the diversity measure is a good idea, but as of now Population relies on the equality operator to remove duplicates, which we should then change? If so, doesn't it make sense to remove the equality operator alltogether, forcing an explicit choice of how to define equality when needed?

I don't think the current equality operator is wrong (except maybe for #315?). It's just that some problem variants have more symmetry resulting in solutions that are virtually the same for the problem at hand, but not in terms of the "equality" operator of the Solution data structure itself. Dealing with that symmetry can be delegated to the diversity measure/Population, since that's the only place it really matters.

**Comment by N-Wouda** [comment]
> Dealing with that symmetry can be delegated to the diversity measure/Population, since that's the only place it really matters.

I agree with this conclusion. The problem might already go away if we have a diversity measure for undirected problems. Let's tackle #101 first and then re-evaluate this issue after.

**Comment by N-Wouda** [comment]
Since we closed #101 I think we can also close this.
